systemtest/src/test/java/io/strimzi/systemtest/operators/topic/TopicReplicasChangeST.java

```diff
@@ -0,0 +1,497 @@
+    @IsolatedTest
+    void testRecoveryOfReplicationChangeDuringCcCrash() {
+        final TestStorage testStorage = new TestStorage(ResourceManager.getTestContext());
+        final int topicPartitions = 3;
+        final int startingTopicReplicationFactor = 2;
+        final int increasedTopicReplicationFactor = 3;
+        final String ccPodName = kubeClient().listPodsByPrefixInName(this.sharedTestStorage.getNamespaceName(), CruiseControlResources.componentName(this.sharedTestStorage.getClusterName())).get(0).getMetadata().getName();
+        final KafkaTopic kafkaTopic = KafkaTopicTemplates.topic(this.sharedTestStorage.getClusterName(), testStorage.getTopicName(), topicPartitions, startingTopicReplicationFactor, 1, this.sharedTestStorage.getNamespaceName()).build();
+
+        // -- 1st stage (start with 2 replicas)
+        resourceManager.createResourceWithWait(kafkaTopic);
+
+        // --- 2nd stage (increasing to 3 replicas)
+        KafkaTopicUtils.waitUntilTopicObservationGenerationIsPresent(this.sharedTestStorage.getNamespaceName(), testStorage.getTopicName());
+        long topicObservationGeneration = KafkaTopicUtils.topicObservationGeneration(this.sharedTestStorage.getNamespaceName(), testStorage.getTopicName());
+
+        KafkaTopicResource.replaceTopicResourceInSpecificNamespace(
+                testStorage.getTopicName(),
+                topic -> topic.getSpec().setReplicas(increasedTopicReplicationFactor), testStorage.getNamespaceName());
+
+        KafkaTopicUtils.waitUntilReplicaChangeResolved(this.sharedTestStorage.getNamespaceName(), testStorage.getTopicName());
+
+        LOGGER.info("ReplicaChange appeared for KafkaTopic: {}/{}. Proceeding to the next stage.", this.sharedTestStorage.getNamespaceName(), testStorage.getTopicName());
+
+        // we should see that KafkaTopic is ongoing
+        KafkaTopicUtils.waitUntilReplicaChangeOngoing(this.sharedTestStorage.getNamespaceName(), testStorage.getTopicName());
+
+        final Map<String, String> ccPod = DeploymentUtils.depSnapshot(this.sharedTestStorage.getNamespaceName(), CruiseControlResources.componentName(this.sharedTestStorage.getClusterName()));
+
+        // --- 3rd stage (during ongoing replicaChange task we induce chaos and delete CruiseControl Pod)
+        kubeClient().deletePodWithName(this.sharedTestStorage.getNamespaceName(), ccPodName);
+
+        // we should see that KafkaTopic is ongoing
+        KafkaTopicUtils.waitUntilReplicaChangeOngoing(this.sharedTestStorage.getNamespaceName(), testStorage.getTopicName());
+
+        // then KafkaTopic replicaChange status should disappear after task is completed by CC
+        KafkaTopicUtils.waitForReplicaChangeStatusNotPresent(this.sharedTestStorage.getNamespaceName(), testStorage.getTopicName());
+
+        // cc should roll
+        DeploymentUtils.waitTillDepHasRolled(this.sharedTestStorage.getNamespaceName(), CruiseControlResources.componentName(this.sharedTestStorage.getClusterName()), 1, ccPod);
```

Comment:
Well, here we are testing that during that crash UTO is handling this flow correctly, i.e., after pod delete we should eventually see ongoing tasks in CC and then such task should disappear.

If I wait for only CC first then I would be unable to check flow of the UTO during recovery.

Reply:
I also don't quite get why you need waitForReplicaChangeStatusNotPresent before CC deletion. You delete CC while the change is ongoing, CC restarts, then the change is eventually retried and completed. What happens is that the TO detects the CC restart, and reset the change status to pending, which enables the retry mechanism.

Reply:
> why you need waitForReplicaChangeStatusNotPresent before CC deletion

`waitForReplicaChangeStatusNotPresent` is invoked after the deletion of CC.

```java
// delete CC pod
kubeClient().deletePodWithName(this.sharedTestStorage.getNamespaceName(), ccPodName);

// check ongoing
KafkaTopicUtils.waitUntilReplicaChangeOngoing(this.sharedTestStorage.getNamespaceName(), testStorage.getTopicName());

 // check not present because this task is already done
 KafkaTopicUtils.waitForReplicaChangeStatusNotPresent(this.sharedTestStorage.getNamespaceName(), testStorage.getTopicName());

// verify that CC deployment rolled
        DeploymentUtils.waitTillDepHasRolled(this.sharedTestStorage.getNamespaceName(), CruiseControlResources.componentName(this.sharedTestStorage.getClusterName()), 1, ccPod);
Member
```